### PR TITLE
Top-N-Sigma: Remove unconditional softmax+sort

### PR DESCRIPTION
**Pull request by TimNN**
## Overview

Currently, the Top-N-Sigma sampler does an unconditional softmax+sort at the end.

In the (common, I believe) case of Top-N-Sigma being followed by Dist, this expensive work is completely wasted.

## Additional information

On my M3 Max MacBook Pro, this PR increases the t/s for `google_gemma-4-E4B-it-Q8_0` by 50%, from ~30t/s to ~45t/s, reducing the time per token by 10ms.

(I'm not sure about the exact API contract between chained samplers and don't know if this might adversely affect other sampler chains that might rely on the current behavior).

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: NO

### CODE REVIEW

**Comment by ddh0** [comment]
I can confirm that this change is correct, at least according to the the original Top-nσ paper ([arXiv:2411.07641](https://arxiv.org/abs/2411.07641)):

> Furthermore, top-nσ operates directly on logits without requiring sorting operations or additional softmax transformations, making it computationally efficient.

---

> In the (common, I believe) case of Top-N-Sigma being followed by Dist, this expensive work is completely wasted.

IIUC, it is _always_ wasted, since in general, each sampler's `_apply` method can make no guarantees about what logits it is receiving; if it needs softmax, it should softmax. Top-nσ does not need it, so I don't think it should apply it.

**Comment by TimNN** [comment]
@ggerganov: The bot (code owners) assigned this to you, and it's been pending for a few weeks.

Could you give an indication of how to proceed? I'm happy to continue waiting, or to just close it if the change isn't as trivial as I think it is.

**Review by ggerganov — APPROVED** [review]
Yes, seems reasonable to remove the softmax here.

**Comment by sunaj** [comment]
I believe we have similar type micro optimization available at: 
llama_sampler_typical_apply (line 1708) where it calls softmax_impl(cur_p, true), changing it to softmax_impl(cur_p, false) is the exact same optimization as the top-n-sigma PR above.
